### DISCUSSION TITLE
One or more namespaces failed to connect

### DISCUSSION THREAD

**Discussion by HichTala** [miguelgrinberg/python-socketio#1308] [Q&A] [answered]
Hi,

I am trying to establish a connection to a server, the server url looks like this: "https://server.url.com/something" so here is my code:

```python
import socketio

sio = socketio.Client(logger=True, engineio_logger=True)
sio.connect('https://server.url.com', transports=['websocket'], namespaces=['/something'])
```

However, I'm encountering the following error:

```
Attempting WebSocket connection to wss://server.url.com/socket.io/?transport=websocket&EIO=4
WebSocket connection accepted with {'sid': 'nIFCUi7Qq3xG195xAAZk', 'upgrades': [], 'pingInterval': 25000, 'pingTimeout': 60000}
Engine.IO connection established
Sending packet MESSAGE data 0/something,{}
Received packet MESSAGE data 0
Namespace / is connected
Received packet MESSAGE data 0/something
Namespace /Zoulpdv is connected
Sending packet MESSAGE data 1
Sending packet MESSAGE data 1/something,
Sending packet CLOSE data None
Engine.IO connection dropped
Exiting write loop task
WebSocket connection was closed, aborting
Waiting for write loop task to end
Exiting read loop task
Traceback (most recent call last):
  File "*******\socket_test.py", line 5, in <module>
    sio.connect('wss://server.url.com', transports=['websocket'], namespaces=['/something'])
  File "*********\lib\site-packages\socketio\client.py", line 163, in connect
    raise exceptions.ConnectionError(
socketio.exceptions.ConnectionError: One or more namespaces failed to connect
```

I am trying to understand if it originates from the server side or if there's something wrong I'm doing (or not doing). what is the cause of the issue, is it in the server side ?  Since I don't have direct access to the server, I'm a bit stuck. Any insights or suggestions on how to troubleshoot this would be greatly appreciated!

Thanks in advance for your help.

**Comment by miguelgrinberg** [comment]
This looks like it's your client ending everything:

```
Sending packet MESSAGE data 1
Sending packet MESSAGE data 1/something,
Sending packet CLOSE data None
```

The "1" packet in the Socket.IO protocol is the `DISCONNECT` packet. The first two lines are disconnecting the `/` and `/something` namespaces, then the third line closes the Engine.IO transport.

Your server is doing something odd, by the way. You are asking for a connection to `/something`, and it is connecting you to `/` and `/something`. Auto-connecting the `/` namespace used to be a requirement in older versions of the Socket.IO protocol, but not anymore. If you are using my Python server then I need to know exactly what version so that I can investigate more.

**Reply by HichTala** [reply]
Thank you for your fast replay, Why the client is disconnecting from the server ? Am I doing something wrong ?

In the server we are not using your Python server. It is a `Socket.io` server, the version is `2.1.4`, basicaly what I want to do is to send an image's path to the server with emiting the event `images`. May be it is just not possible with this library or may be should I do something differently ?

**Reply by miguelgrinberg** [reply]
Without seeing your client I cannot answer. The log suggests you are disconnecting explicitly, so this is likely a bug in your code.

**Reply by HichTala** [reply]
I gave the code I am executing 

```python
import socketio

sio = socketio.Client(logger=True, engineio_logger=True)
sio.connect('https://server.url.com', transports=['websocket'], namespaces=['/something'])
```
I am not executing anyhting else...

**Reply by HichTala** [reply]
I found that if I delete the `namespaces` option, the error message disappears, maybe it's related.

**Reply by HichTala** [reply]
it looks like it is related to this part of code 
https://github.com/miguelgrinberg/python-socketio/blob/e8bab71413f9a70a089135f82ef0a87473982de7/src/socketio/client.py#L157-L160

`self._connect_event.clear()` is making `self._connect_event.wait(timeout=wait_timeout)` returns `True`, so `while` loop is exited, and `set(self.namespaces)`, `set(self.connection_namespaces)` have different values.

I don't know if they should be different at this stage or if there is a probleme before...

**Reply by HichTala** [reply, marked as answer]
Adding `'/'` to the namspaces list reslov the issue, 

```python
import socketio

sio = socketio.Client(logger=True, engineio_logger=True)
sio.connect('https://server.url.com', transports=['websocket'], namespaces=['/', '/something'])
```

The problem may be that the client connects to `'/'` when you don't think it should. One straight forward solution will be to add `/` to `self.connection_namespaces` in the same time it is added to `self.namespace` but I can't find where `self.namespace` is modified and I don't have time to investigate more today, but I can open a pull request in the future to look at it if you want.

**Reply by miguelgrinberg** [reply]
Thanks for the detailed analysis. It looks like you are asking for one namespace, and the server is connecting you to two, which confuses the client. I'll see if I can reproduce this and add better handling.

Update: never mind, I think I figured it out. You said you are using Socket.IO 2.x correct? Check the version compatibility table, that is a very old version of the JS server, you need to downgrade your Python if you want to use such an old version. Or else, upgrade to a 4.x on the JS side and let me know if there is an improvement.

**Reply by HichTala** [reply]
Okay, it's clearer now, thanks. I'll check with the server administrator to see if it's possible to update the Socket.io version.
